cli/medperf/comms/auth/auth0.py

```diff
@@ -0,0 +1,247 @@
+        backend using the device authorization flow.
+
+        """
+        device_code_response = self.__request_device_code()
+
+        device_code = device_code_response["device_code"]
+        user_code = device_code_response["user_code"]
+        verification_uri_complete = device_code_response["verification_uri_complete"]
+        interval = device_code_response["interval"]
+
+        config.ui.print(
+            "Please go to the following link to complete your login request:\n"
+            f"\t{verification_uri_complete}\n\n"
+            "Make sure that you will be presented with the following code:\n"
+            f"\t{user_code}\n\n"
+        )
+        token_response, issued_at = self.__get_device_access_token(
+            device_code, interval
+        )
+        access_token = token_response["access_token"]
+        refresh_token = token_response["refresh_token"]
+        expires_in = token_response["expires_in"]
+
+        set_credentials(
+            {
+                "access_token": access_token,
+                "refresh_token": refresh_token,
+                "expires_in": expires_in,
+                "issued_at": issued_at,
+            }
+        )
+
+    def __request_device_code(self):
+        """Get a device code from the auth0 backend to be used for the authorization process"""
+        url = f"https://{self.domain}/oauth/device/code"
+        headers = {"content-type": "application/x-www-form-urlencoded"}
+        body = {
+            "client_id": self.client_id,
+            "audience": self.audience,
+            "scope": "offline_access",  # to get a refresh token
+        }
+        res = requests.post(url=url, headers=headers, data=body)
+
+        if res.status_code != 200:
+            self.__raise_errors(res, "Login")
+
+        return res.json()
+
+    def __get_device_access_token(self, device_code, polling_interval):
+        """Get the access token from the auth0 backend associated with
+        the device code requested before. This function will keep polling
+        the access token until the user completes the browser flow part
+        of the authorization process.
+
+        Args:
+            device_code (str): A temporary device code requested by `__request_device_code`
+            polling_interval (float): number of seconds to wait between each two polling requests
+
+        Returns:
+            json_res (dict): the response of the successful request, containg the access/refresh tokens pair
+            issued_at (float): the timestamp when the access token was issued
+        """
+        url = f"https://{self.domain}/oauth/token"
+        headers = {"content-type": "application/x-www-form-urlencoded"}
+        body = {
+            "grant_type": "urn:ietf:params:oauth:grant-type:device_code",
+            "device_code": device_code,
+            "client_id": self.client_id,
+        }
+
+        while True:
+            time.sleep(polling_interval)
+            issued_at = time.time()
```

Comment:
Silly question: should this be before or after post?

Reply:
I chose to record before, based on "better to refresh the token early than being late".